dodiscover/toporder/_base.py

```diff
@@ -0,0 +1,629 @@
+                if ((not G_excluded.has_edge(p, c)) and A_dense[p, c] == 1) or G_included.has_edge(
+                    p, c
+                ):
+                    pot_parents.append(p)
+            if len(pot_parents) > 0:
+                parents = self._variable_selection(X[:, pot_parents], X[:, c], pot_parents, c)
+                for parent in parents:
+                    A[parent, c] = 1
+
+        return A
+
+    def exclude_edges(self, A: NDArray) -> NDArray:
+        """Update `self.context` excluding edges not admitted in `A`.
+
+        Parameters
+        ----------
+        A : np.ndarray
+            Adjacency matrix representation of an arbitrary graph (directe or undirected).
+            If `A[i, j]`, add edge (i, j) to `self.context.excluded_edges`.
+        """
+        # self.context.excluded_edges: nx.Graph with excluded edges
+        d = A.shape[0]
+        for i in range(d):
+            for j in range(d):
+                if i != j and A[i, j] == 0 and (not self.context.included_edges.has_edge(i, j)):
+                    self.context.excluded_edges.add_edge(i, j)
+
+    def pns(self, A: NDArray, X: NDArray) -> NDArray:
+        """Preliminary Neighbors Selection (PNS) pruning on adjacency matrix `A`.
+
+        Variable selection preliminary to CAM pruning.
+        PNS :footcite:`Buhlmann2013` allows to scale CAM pruning to large graphs
+        (~20 or more nodes), with sensitive reduction of computational time.
+
+        Parameters
+        ----------
+        A : np.ndarray
+            Adjacency matrix representation of a dense graph.
+        X : np.ndarray of shape (n_samples, n_nodes)
+            Dataset with observations of the causal variables.
+
+        Returns
+        -------
+        A : np.ndarray
+            Pruned adjacency matrix.
+
+        References
+        ----------
+        .. footbibliography::
+        """
+        num_nodes = X.shape[1]
+        for node in range(num_nodes):
+            X_copy = np.copy(X)
+            X_copy[:, node] = 0
+            reg = ExtraTreesRegressor(n_estimators=500)
+            reg = reg.fit(X_copy, X[:, node])
+            selected_reg = SelectFromModel(
+                reg,
+                threshold="{}*mean".format(self.pns_threshold),
+                prefit=True,
+                max_features=self.pns_num_neighbors,
+            )
+            mask_selected = selected_reg.get_support(indices=False)
+
+            mask_selected = mask_selected.astype(A.dtype)
+            A[:, node] *= mask_selected
+
+        return A
+
+    def _dag_check_included_edges(self) -> None:
+        """Check that the edges included in `self.context` does not violate DAG condition."""
+        is_dag = nx.is_directed_acyclic_graph(self.context.included_edges)
+        if nx.is_empty(self.context.included_edges):
+            is_dag = True
+        if not is_dag:
+            raise ValueError("Edges included in the graph violate the acyclicity condition!")
```

Comment:
Yes, and it works a bit differently for CAM with respect to the other methods.

- CAM infers the topological ordering by explicitly building a fully connected DAG one edge at the time, since this has a 1:1 mapping with the ordering. While inferring the topological order it uses a binary matrix `directed_paths` of shape `(num,_nodes, num_nodes)` that stores directed paths  between nodes, i.e. `directed_paths[i, j]=1` means that there is a directed path from node `i` to node `j`. In such case, it will simply forbid edges that allow for a cyclic path from `j` to `i` to be added. So, when an edge is included as an external constraint, I simply update the `directed_paths` matrix accordingly
- The other methods infer the topological order vector directly, one leaf at the time (rather then building a fully connected matrix). So, let's say a  user includes `i -> j` as an edge: when selecting leaf nodes during the ordering, I forbid `i` to be picked as a leaf before `j`.  

I hope this clarifies a bit, feel free to ask more.

Reply:
I would add this into the `Notes` section of the class to clarify what sort of "priori" knowledge is supported and perhaps also worth mentioning what isn't possible to encode (if any). In the actual `plot_score_alg.py` example, if there's an easy way to demonstrate this usage of apriori knowledge, or having a very very short 2nd example that just demonstrates this, then that would be great!

For context, we're trying to make sure apriori knowledge is usable when it is provided, but the challenge is being very explicit about this to users, otw it's not clear when/where apriori causal knowledge is usable.

Reply:
Ok, I will add to the `Notes` that it is possible to encode, meaning specific edges, vs what is not possible, meaning constraints in the order. As we discussed I would add the latter in a separate PR.

For included edges encoding I can add an additional example, do you reccomend to have a separate script or it is fine to have it on top of `plot_score_alg.py`?

Reply:
Is it okay to have it as a separate script to keep this example nice and concise? We can add a reference link between each examples, so they're easy to find. The separate script can be super short and just go straight to demonstrating what sort of prior knowledge is needed. Then link back to this example for anyone wondering what the methods are.

Reply:
I added the example to illustrate how to encode priori knowledge for the order-based methods.
See https://github.com/py-why/dodiscover/pull/129/files#diff-c2f2f9f69d6a4fac3ef5d0b96cfc57eb8769f2d7a238a430d328cbeef6d80a58.

I show how to encode both included and excluded edges and plot the graph to illustrate the effect.